zipline/utils/events.py

```diff
@@ -522,19 +501,43 @@ def get_last_trading_day_of_week(dt, env):
         dt = env.next_trading_day(dt)
         # Traverse forward until we hit a week border, then jump back to the
         # previous trading day.
-        while dt.date().weekday() > prev.date().weekday():
+        while dt and dt.date().weekday() > prev.date().weekday():
             prev = dt
             dt = env.next_trading_day(dt)
 
         if env.is_trading_day(prev):
-            return prev.date()
+            return prev
         else:
-            return env.previous_trading_day(prev).date()
+            return env.previous_trading_day(prev)
 
     date_func = get_last_trading_day_of_week
 
 
-class NthTradingDayOfMonth(StatelessRule):
+class TradingDayOfMonthRule(six.with_metaclass(ABCMeta, StatelessRule)):
+
+    def should_trigger(self, dt, env):
+        return self.get_trigger_day_of_month(dt, env) == env.normalize_date(dt)
+
+    @abstractmethod
+    def date_func(self, dt, env):
+        raise NotImplementedError
+
+    def get_trigger_day_of_month(self, dt, env):
+        if self.month == dt.month:
+            # We already computed the day for this month.
+            return self.date
+
+        self.date = self.date_func(dt, env)
+        if self.td_delta and self.date:
+            try:
+                self.date = env.add_trading_days(self.td_delta, self.date)
+            except NoFurtherDataError:
+                self.date = None
+
+        return self.date
```

Comment:
If `self.month`, `self.td_delta`, and `self.date` are expected in all subclasses, you might move setting them up to the `__init__` of this base class.